### PR TITLE
Support any encoding for fileview

### PR DESCRIPTION
**Pull request by schadin**
Changes proposed in this pull request:
-
Added support all available encodings .NET:
- in settings
- in file viewer
 
Screenshots before and after (if PR changes UI):
-
List of encodings
encodings_before

encodings

How did I test this code:
 - 
Manual testing: viewing files in different encodings

Has been tested on:
-
 - GIT 2.12.2
 - Windows 10

### CODE REVIEW

File: GitUI/Editor/FileViewer.cs [hunk 1]
```diff
@@ -71,11 +71,11 @@ public FileViewer()
 
             IsReadOnly = true;
 
-            this.encodingToolStripComboBox.Items.AddRange(new Object[]
-                                                    {
-                                                        "Default (" + Encoding.Default.HeaderName + ")","DOS852", "ASCII",
-                                                        "Unicode", "UTF7", "UTF8", "UTF32"
-                                                    });
+            var encodingArray = (from e in AppSettings.AvailableEncodings.Values
+                                select e.EncodingName).ToArray();
```
Comment (RussKie) [inline GitUI/Editor/FileViewer.cs]: this should be sorted in some meaningful manner.
maybe even group by dos, windows, mac, etc

```
--- DOS ---
encoding1 (DOS)
encoding2 (DOS)
-- Windows ---
encoding1 (Win)
encoding2 (Win)
```

File: GitCommands/Settings/AppSettings.cs [hunk 2]
```diff
@@ -1179,13 +1179,11 @@ public static void SaveSettings()
...
-            addEncoding(new ASCIIEncoding());
-            addEncoding(new UnicodeEncoding());
-            addEncoding(new UTF7Encoding());
-            addEncoding(new UTF8Encoding(false));
-            addEncoding(System.Text.Encoding.GetEncoding("CP852"));
-
+            foreach(var enc in Encoding.GetEncodings())
+            {
+                addEncoding(enc.GetEncoding());
+            }
```
Comment (RussKie) [inline GitCommands/Settings/AppSettings.cs]: what is the use-case to have all of them?

I don't think it is a good idea to overload the user with the unnecessary data. Would it perhaps be better to allow the user to toggle all encoding, if some quirky encoding is required? 
And by default the user would continue to see the selected list of encodings.

File: GitCommands/Settings/AppSettings.cs [hunk 3]
```diff
@@ -1179,13 +1179,11 @@ public static void SaveSettings()
         public static void LoadSettings()
         {
             Action<Encoding> addEncoding = delegate(Encoding e) { AvailableEncodings[e.HeaderName] = e; };
-            addEncoding(Encoding.Default);
-            addEncoding(new ASCIIEncoding());
-            addEncoding(new UnicodeEncoding());
-            addEncoding(new UTF7Encoding());
-            addEncoding(new UTF8Encoding(false));
```
Comment (jbialobr) [inline GitCommands/Settings/AppSettings.cs]: Make sure that the UTF8 without BOM encoding is used.

File: GitCommands/Settings/AppSettings.cs (hunk 2 above)
Comment (schadin) [inline GitCommands/Settings/AppSettings.cs]: Good day! We needed cp866, but add a single character encoding seemed to me unfortunate decision. I'll think about what can be done in a convenient way.

Comment (jbialobr) [inline GitCommands/Settings/AppSettings.cs]: Maybe sort them all alphabetically. The user usually knows the encoding name that is seeking for.

File: GitCommands/Settings/AppSettings.cs (hunk 3 above)
Comment (schadin) [inline GitCommands/Settings/AppSettings.cs]: I think about grouping like notepad++.

File: GitCommands/Settings/AppSettings.cs (hunk 2 above)
Comment (schadin) [inline GitCommands/Settings/AppSettings.cs]: I think about grouping like notepad++.

**Comment by schadin** [comment]
It's bad pull request. i'm sorry.

**Comment by RussKie** [comment]
@schadin, nothing to be sorry about. :)
It is not a bad pr, but it can certainly be improved.

**Comment by vbjay** [comment]
What about an auto complete textbox instead?

**Comment by schadin** [comment]
I want to create a new form with a list of available encodings. Users will be able to choose the encoding independently.

**Comment by KindDragon** [comment]
I think better if have in list only with: ANSI (Windows), ASCII (DOS) and Unicode/UTF encodings, but allow user in settings dialog change Windows and DOS encoding

**Pull request closed without merging**